### DISCUSSION TITLE
DB locks  on postal.queued_messages table

### DISCUSSION THREAD

**Discussion by onurbektas** [postalserver/postal#2620] [Help with using Postal]
Hi, 

We have two postal servers using the same database. In order to increase the number of messages sent, we increased the number of workers. After that, we observed db lock errors on postal logs.
 
`postal-worker-2    | [8] [2023-09-19T09:18:27.625] WARN -- : [UO3YL2RXSR] ActiveRecord::LockWaitTimeout: Mysql2::Error::TimeoutError: Lock wait timeout exceeded; try restarting transaction: UPDATE `queued_messages` SET `queued_messages`.`locked_by` = 'host:mx2 pid:8 job:UO3YL2RXSR', `queued_messages`.`locked_at` = '2023-09-13 09:17:37.312956' WHERE (retry_after IS NULL OR retry_after <= '2023-09-19 09:18:07.313006') AND `queued_messages`.`batch_key` = 'outgoing-gmail.com' AND `queued_messages`.`ip_address_id` = 93 AND `queued_messages`.`locked_by` IS NULL AND `queued_messages`.`locked_at` IS NULL LIMIT 100
     `

After analyzing the issue, we found that  postal.queued_messages  table has too many locks 

![image](https://github.com/postalserver/postal/assets/3854947/65717f70-2039-4536-a93d-14e47004a4aa)

show proccesslist command output shows the details of locked operations waiting 
![image](https://github.com/postalserver/postal/assets/3854947/6da7578e-386f-4513-af0b-2e6e435cc1bc)

Is there any solution for this problem ?

**Comment by willpower232** [comment]
Is it possible that one of your servers had an awkward restart and didn't clear their database connection properly?

It looks like you might need to kill one of the processes you listed in order to free the others.

If you're able to free the locks, does the problem re-occur?

**Reply by onurbektas** [reply]
Hi, 

**Is it possible that one of your servers had an awkward restart and didn't clear their database connection properly?**
Running postal on a single server gives the same result in our case.  Shutting down the db and postal and starting on postal on single server reproduced the issue. 

**It looks like you might need to kill one of the processes you listed in order to free the others.**
which one should we kill, how do we choose ?

**If you're able to free the locks, does the problem re-occur?**
Problem occurs under load. by load I mean around 150K mail per hour. After all the mails  have been sent locks are freed automatically. 

So far, I increased the number read/write of I/O threads  and innodb io capacity  (innodb_write_io_threads=8, innodb_read_io_threads=6  and  innodb_io_capacity = 350) thinking that it might be a database performance issue, but it does not change the results. I/O operations on disk are around 3K write and 0.5K read under load and they are within performance limit as system runs on SSD. 

I also increase innodb_lock_wait_timeout to 120sn (default 60), that does not help as well. Lastly, as we increase the worker count,  number of locks increases.

**Comment by willpower232** [comment]
> Lastly, as we increase the worker count, number of locks increases

I should have asked how many workers you are using

> After all the mails have been sent locks are freed automatically

So all the messages end up sent eventually?

**Comment by onurbektas** [comment]
> > Lastly, as we increase the worker count, number of locks increases
> 
> I should have asked how many workers you are using
> 
Tried using 6,8,10,16 workers, nothing changed. 
> > After all the mails have been sent locks are freed automatically
> 
> So all the messages end up sent eventually?

Yes, all the messages sent without any problem.

After digging the sql , we found that we are hitting the  max table size for postal.queued_messages table. Adding following variables  to my.cnf relieves the problem. Now we are getting less lock errors.  

[mysqld]
max_heap_table_size = 2G
tmp_table_size = 2G

To completely get rid of the error (for testing purposes ) we also change the db engine for the table  with, 

`alter table queued_messages ENGINE=MEMORY ;`

this completely  eliminate the locking in db. However, number of messages send is still around 2000 per minute.  We thought that hitting the limit was related with the locking in db,  but it seems it is not.  Still investigating..

**Reply by willpower232** [reply]
Thanks for your efforts, good to know that the messages are all sent eventually so it should be fine, just weird in the short term.

Hope you're able to figure it out and looking forward to whatever the answer is!
